pghoard/basebackup/base.py

```diff
@@ -792,6 +796,23 @@ def get_backup_end_segment_and_time(self, db_conn):
             db_conn.commit()
             return None, backup_end_time
 
+        if self.pg_version_server >= 170000:
+            # In PostgreSQL versions prior to 17, the `pg_walfile_name` function could return the previous WAL file
+            # if the provided LSN fell at the boundary between WAL segments (e.g., at the start of the next WAL file).
+            # This could lead to inconsistencies in the backup metadata, as the LSN obtained from `pg_backup_stop`
+            # might not correspond to the actual WAL file that contains the backup's end LSN.
+            #
+            # To handle this, we had to fetch the current LSN after stopping the backup and use it to determine the
+            # correct WAL file name, ensuring the backup's end segment was accurately identified.
+            #
+            # However, in PostgreSQL 17 and later, this issue is resolved. `pg_walfile_name` now always returns
+            # the current WAL file, regardless of where the LSN falls within the segment. As a result, the extra step
+            # of manually fetching the LSN after `pg_backup_stop` is no longer necessary— the output from
+            # `pg_backup_stop` is now accurate and can be relied upon directly.
+
+            cursor.execute(f"SELECT pg_walfile_name('{backup_stop_result.lsn}')")
+            return cursor.fetchone()[0], backup_stop_result.backup_end_time
```

Comment:
I don't think the logic for PG < 17 makes sense. 

The backupstop LSN is the one returned by XLogInsert upon inserting the backup stop record, which is the end of the record. 
So if this is at a boundary, and the walfilename is the previous one it's actually fine because what we care about is before that LSN.

Reply:
Ah I just noticed an error on my comment:
```
# This could lead to inconsistencies in the backup metadata, as the LSN obtained from `pg_backup_stop`
```
Should be
```
# This could lead to inconsistencies in the backup metadata, as the WAL file obtained from `pg_walfile_name(pg_current_wal_lsn())`
```

Anyways, using the backupstop LSN for PG < 17 does make sense but at the same time I am a bit afraid of removing the old behavior, because maybe we are not considering something.